### PR TITLE
OLED Extension!

### PR DESCRIPTION
**Pull request by ryan-berwald**
I2C SSD1306 OLED Extension complete with image, text, and Active Layer support. Looking for guidance on how to get this merged! Thanks!

### CODE REVIEW

File: docs/oled.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+## Setup
+
+1. Find a BMP image you like and convert it to an indexed BMP by following the above instructions
+2. Place image somewhere on drive and make a note of the absolute path for later.
+3. Enable extension by something similar to the code below, inputting image path or text where necessary
+
+```
+from from kmk.extensions.oledDisplay import oled
+[...KMK CODE ]
```
Comment (kdb424) [inline docs/oled.md:64]: We try to keep consistent and use
\`\`\`python

This gives syntax highlighting.

File: docs/oled.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+keyboard.extensions.append(oled(board.GP21, board.GP20, toDisplay = "IMAGE\\PATH", oWidth = OLEDWIDTH, oHeight = OLEDHEIGHT, tileWidth = WIDTH OF IMAGE, tileHeight = HEIGHT OF IMAGE, gridWidth = IMAGE REPETITION ACROSS, gridHeight = IMAGE REPETITION UP))
+# Note, my image is a 128x32 indexed bmp, if I had a 32x32 indexed BMP, it can repeat across my screen 4 times, so my constructor would look like this instead:
+# keyboard.extensions.append(oled(board.GP21, board.GP20, toDisplay = "images\\f.bmp", oWidth = 128, oHeight = 32, tileWidth = 32, tileHeight = 32, gridWidth = 4, gridHeight = 1))
+[...KMK CODE]
+```
```
Comment (kdb424) [inline docs/oled.md:69]: here as well

File: docs/oled.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+B. Text
+
+```
```
Comment (kdb424) [inline docs/oled.md:73]: and here

File: docs/oled.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+from from kmk.extensions.oledDisplay import oled
+[...KMK CODE ]
+keyboard.extensions.append(oled(board.GP21, board.GP20, "Hello World"))
+[...KMK CODE]
+```
+
+C. Active Layer
+
+```
```
Comment (kdb424) [inline docs/oled.md:82]: and here

File: docs/oled.md (hunk 3 above)
Comment (kdb424) [inline docs/oled.md:73]: and here

**Comment by kdb424** [comment]
OLED's are a tough one to really "support" directly as they can do so much. I love the basis of this. I'll give it a deeper look, but it's a great start to a framework to let users do as they please, which is the goal for something that's "useless" but also fun like OLEDs. Maybe we split this into a library with useful things you *can* do with the oleds, and an extension/module that gets access. Just something to consider, though I don't have a solid plan on how to do that. Thanks!

File: docs/oled.md (hunk 1 above)
Comment (ryan-berwald) [reply docs/oled.md:64]: Added python tag to the code blocks!

**Comment by ryan-berwald** [comment]
> 
> 
> OLED's are a tough one to really "support" directly as they can do so much. I love the basis of this. I'll give it a deeper look, but it's a great start to a framework to let users do as they please, which is the goal for something that's "useless" but also fun like OLEDs. Maybe we split this into a library with useful things you _can_ do with the oleds, and an extension/module that gets access. Just something to consider, though I don't have a solid plan on how to do that. Thanks!

Yes! I totally agree and am on board with that. Ultimately, for my current use case (a macropad), I found it to be particularly useful to be able to show your active layer, or if you wanted to have it be more decorative, I figured it may be nice to throw a bmp on there. I did have some other ideas that could allow those to be used simultaneously but those will take a good bit more work, and I thought this was a good base for oleds. Like you said, this at the very least could let people build on top of an existing base instead of having to build their own OLED lib. 

Also, I fixed the docs!

**Comment by kdb424** [comment]
If you are personally interested in splitting this out into a library, I'll wait for a proper review. If you are happy with the state of your work, if you mark the docs "WIP", and add "example" to the module name, and run `make fix-formatting fix-isort` and push those changes up, the test will pass, and I'll get it merged.

**Comment by ryan-berwald** [comment]
> 
> 
> If you are personally interested in splitting this out into a library, I'll wait for a proper review. If you are happy with the state of your work, if you mark the docs "WIP", and add "example" to the module name, and run `make fix-formatting fix-isort` and push those changes up, the test will pass, and I'll get it merged.

Alright, sorry for the 2 month delay on this. So with you being one of the major maintainers of KMK, what would be ideal for the "creative direction" of this for lack of a better term:

1. A library that would accept a callback function with **kwargs and you essentially would extend it yourself.
          ex. Oled class instantiates the display itself, runs a callback function with kwargs provided by user. Callback function would               have to be written by the user if they wanted something outside of BMPs and Active Layer which I can provide.

- In regards to this, I have BMP and text already written and basically ready to go, just more extensible for the future.

2. Leave it as is with BMP and Active Layer/Text as your options

Ultimately, option 1 would be extensible but less user-friendly. Option 2 is obviously more user friendly but is more heavy weight due to built in functionality.

**Comment by kdb424** [comment]
Quite honestly, anyone wanting extreme customization will likely write their own from scratch, similar to OLED on other keyboard firmwares. Option 2 seems great for most users that just want something that works.

**Comment by ryan-berwald** [comment]
When running the make command I am getting an error that I have no idea how to resolve.


sync: unrecognized option '--dev'
make: *** [Makefile:49: .devdeps] Error 1

**Comment by kdb424** [comment]
> When running the make command I am getting an error that I have no idea how to resolve.
> 
> sync: unrecognized option '--dev'
> make: *** [Makefile:49: .devdeps] Error 1

You'll have to have pipenv installed I believe. Can open the makefile to see what's actually happening.

**Comment by ryan-berwald** [comment]
The only way I'm aware of submitting something as WIP is doing it through another PR with the docs file separate so I changed the name to WIP for the time being. Do you want me to submit a second PR with the doc as a draft PR?

**Comment by kdb424** [comment]
> The only way I'm aware of submitting something as WIP is doing it through another PR with the docs file separate so I changed the name to WIP for the time being. Do you want me to submit a second PR with the doc as a draft PR?

Hit the "convert to draft" button on github.

**Comment by ryan-berwald** [comment]
Okay, this is now set as a draft.

**Comment by kdb424** [comment]
Closing due to it being stale.

**Pull request closed without merging**